subsys/bluetooth/mesh/sensor_types.c

```diff
@@ -0,0 +1,952 @@
+
+		if (repr->flags & (HAS_HIGHER_THAN | HAS_INVALID)) {
+			raw = type_max - 2;
+		} else if (repr->flags & HAS_UNDEFINED) {
+			raw = type_max - 1;
+		} else {
+			return -ERANGE;
+		}
+	}
+
+	switch (format->size) {
+	case 1:
+		net_buf_simple_add_u8(buf, raw);
+		break;
+	case 2:
+		net_buf_simple_add_le16(buf, raw);
+		break;
+	case 3:
+		net_buf_simple_add_le24(buf, raw);
+		break;
+	case 4:
+		net_buf_simple_add_le32(buf, raw);
+		break;
+	default:
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int scalar_decode(const struct bt_mesh_sensor_format *format,
+			 struct net_buf_simple *buf, struct sensor_value *val)
+{
+	const struct scalar_repr *repr = format->user_data;
+
+	if (buf->len < format->size) {
+		return -ENOMEM;
+	}
+
+	s32_t raw;
+
+	switch (format->size) {
+	case 1:
+		if (repr->flags & SIGNED) {
+			raw = (s8_t) net_buf_simple_pull_u8(buf);
+		} else {
+			raw = net_buf_simple_pull_u8(buf);
+		}
+		break;
+	case 2:
+		if (repr->flags & SIGNED) {
+			raw = (s16_t) net_buf_simple_pull_le16(buf);
+		} else {
+			raw = net_buf_simple_pull_le16(buf);
+		}
+		break;
+	case 3:
+		raw = net_buf_simple_pull_le24(buf);
+
+		if ((repr->flags & SIGNED) && (raw & BIT(24))) {
+			raw |= (BIT_MASK(8) << 24);
+		}
+		break;
+	case 4:
+		raw = net_buf_simple_pull_le32(buf);
+		break;
+	default:
+		return -ERANGE;
+	}
+
+	u32_t max_value = scalar_max(format);
+	s32_t min_value = scalar_min(format);
+
+	if (raw < min_value || raw > max_value) {
+		return -ERANGE;
+	}
+
+	s64_t million = mul_scalar(raw * 1000000LL, repr);
+
+	val->val1 = million / 1000000LL;
+	val->val2 = million % 1000000LL;
+
+	return 0;
+}
+
+static int boolean_encode(const struct bt_mesh_sensor_format *format,
+			  struct net_buf_simple *buf,
+			  const struct sensor_value *val)
+{
+	if (net_buf_simple_tailroom(buf) < 1) {
+		return -ENOMEM;
+	}
+
+	net_buf_simple_add_u8(buf, !!val->val1);
+	return 0;
+}
+
+static int boolean_decode(const struct bt_mesh_sensor_format *format,
+			  struct net_buf_simple *buf, struct sensor_value *val)
+{
+	if (buf->len < 1) {
+		return -ENOMEM;
+	}
+
+	u8_t b = net_buf_simple_pull_u8(buf);
+
+	if (b > 1) {
+		return -EINVAL;
+	}
+
+	val->val1 = b;
+	val->val2 = 0;
+	return 0;
+}
+
+static int exp_1_1_encode(const struct bt_mesh_sensor_format *format,
+			  struct net_buf_simple *buf,
+			  const struct sensor_value *val)
+{
+	if (net_buf_simple_tailroom(buf) < 1) {
+		return -ENOMEM;
+	}
+
+	net_buf_simple_add_u8(buf, sensor_powtime_encode((val->val1 * 1000) +
+							 (val->val2 / 1000)));
+	return 0;
+}
+
+static int exp_1_1_decode(const struct bt_mesh_sensor_format *format,
+			  struct net_buf_simple *buf, struct sensor_value *val)
+{
+	if (buf->len < 1) {
+		return -ENOMEM;
+	}
+
+	u64_t time_ns = sensor_powtime_decode_ns(net_buf_simple_pull_u8(buf));
+
+	val->val1 = time_ns / 1000000L;
+	val->val2 = time_ns % 1000000L;
+	return 0;
+}
+
+static int float32_encode(const struct bt_mesh_sensor_format *format,
```

Comment:
This is common for many GATT services. I would be very happy if this was upstreamed, could be used in peripheral_ht and central_ht samples. Or moved to somewhere that GATT services can access them as well

Reply:
I see. In fact, everything in this file is just encoding and decoding of characteristics values. I considered making a more generic framework, but I wanted to keep the `sensor_value` representation for this module. It might be worth looking into making this a generic module that the GATT services can use as well.

Reply:
Yeah, did you make a task for this as well? So that we don't forget?

Reply:
yep